dialer_test.go

```diff
@@ -5,10 +5,19 @@ import (
 	"crypto/x509"
 	"io/ioutil"
 	"reflect"
-	"runtime"
 	"testing"
 )
 
+// funcEquals checks if the two given functions are the same. This grabs the
+// function pointer of each of them and compares them. This seems to be a viable
+// way to work around the fact that Go does not have actual function equality.
+func funcEquals(f1, f2 interface{}) bool {
+	av := reflect.ValueOf(f1).Pointer()
+	bv := reflect.ValueOf(f2).Pointer()
+
+	return av == bv
```

Comment:
The internet tells me pointer equality is not necessarily safe as the compiler might choose to inline funca into funcb and then the pointer comparison will return true when comparing funca == funcb.

That said, there's zero chance of that occurring in this case so I'm not sure we need to worry about it? Maybe the comment should note that this is ostensibly safe if and only if the functions being compared are guaranteed to be uncoupled?